### PR TITLE
feat(404 Page): Create custom 404 page

### LINKED ISSUE #4: 404
We need a 404 page.

### PR DESCRIPTION
**Pull request by amberwelby**
# Screenshot
Screen 1: when navigating from a link within the site
image

Screen 2: when navigating from a link from outside of the site (or typed)
image

# Description
Adding a set of error pages, notably a generic and 404 page, to cover as many potential issues as possible. Depending on the error, users may return home or report the issue on GitHub.

### CODE REVIEW

**Review by Kara-Zor-El — DISMISSED** [review]
Code seems fine but I think maybe let's have design mock something up

**Comment by spotandjake** [comment]
closes: #4

File: styles/Error.module.scss
Comment (Kara-Zor-El) [inline styles/Error.module.scss]: Can you add #222 to the typescript config as a color (feel free to message on discord if you need more explanation)

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: looking here https://nextjs.org/docs/pages/building-your-application/routing/custom-error#more-advanced-error-page-customizing, this page can appear in more cases then the `404` can we change this to in the `404` case show this message otherwise can we just say `A client side error occured`. with a button back to the homepage.

**Review by spotandjake — COMMENTED** [review]
One small nit otherwise everything looks great.

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: Woah why is this an `any`?

can you also break the type definiton into a seperate interface like:
```ts
interface Props {
  statusCode: any; // replace any with whatever type it is meant to be, probably `number` but it might say on the next docs
}
```

This just makes the code more readable.

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: Mind just adding a comment above this variable explaining its role, also just thinking maybe `returnHref` is a better name just so its clear it is a url. (The naming is minor if you don't agree feel free to leave it).

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: can you make this a `main` instead of a `div` its the main page content, it also follows the `homepage`.

File: pages/_error.tsx
Comment (Kara-Zor-El) [inline pages/_error.tsx]: I think having text similar to your first draft rather than saying that "a 404 error occurred on the server". Something that tells a non tech savvy person that the page doesn't exist or is being developed

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: instead of using a class here maybe use css nesting something like:
```scss
.container {
  & > h1 {} // styling
}
```
The reason being we should enforce styles based off semantics and classes should add more context if they do not add any context why do we have the class hope that makes sense.

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: Same note as above.

File: pages/_error.tsx
Comment (spotandjake) [reply pages/_error.tsx]: I noticed below in the body you do `statusCode ? '${statusCode}' : 'default err'`, so maybe it should be `statusCode?: any` but again replace `any` with whatever its meant to be.

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: Is there a reason this is in a div? just curious.

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]:
```suggestion
              Return Home
```
More context ussualy helps accessibility.

**Review by Kara-Zor-El — COMMENTED** [review]
Additionally, I noticed that on mobile sizes, it is different. Happy to leave that and come back later (I am doing that for the main page) but thought I'd just make a note of it now
image

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]:
```suggestion
              href={returnTo}
              prefetch={true}
```
Double check thats the right prop, but the only place the user will be navigating from this page is home or the back button so it probably makes sense to prefetch the home page here, can you test it and if it doesnt incur any major wait or load time enable the prefetching?

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: Why are these `any`?

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: small nit can we change the name `canvas` it can be a bit confusing given there is an html canvas element. maybe something make this `container` and change the container class to be `content` or `pageContent`.

**Review by spotandjake — COMMENTED** [review]
Mostly looks great Amber, I did have a few notes some on types we should try and use strong typing where possible any allows for bugs to slip through and makes typescript almost useless.

There are just a few on styling and html semantics, they are minor.

great job overall

File: pages/_error.tsx
Comment (amberwelby) [reply pages/_error.tsx]: Are we thinking the first draft from the description (longer and personable) or the last draft (sort and simpler)?

File: pages/_error.tsx
Comment (amberwelby) [reply pages/_error.tsx]: Gave me some specific control over spacing, I'll try to do that without the div though

**Comment by amberwelby** [comment]
> Additionally, I noticed that on mobile sizes, it is different. Happy to leave that and come back later (I am doing that for the main page) but thought I'd just make a note of it now image

I noticed that too, it starts about 410 w and I don't know why the behaviour changes there

File: pages/_error.tsx
Comment (amberwelby) [reply pages/_error.tsx]: Doesn't seem to make a big difference in timing either way, let me know if you want it there or not. There is a pretty jarring white screen before the 404 loads, but I don't remember if it's always done that or not

File: pages/layout.tsx
Comment (Kara-Zor-El) [inline pages/layout.tsx:1]: move to `layout/ErrorLayout.tsx` or something along those lines

Comment (spotandjake) [inline pages/layout.tsx:1]: This is going to need to move to `layout/Eror.tsx`

File: pages/layout.tsx
Comment (Kara-Zor-El) [inline pages/layout.tsx]: if were not passing in any props, why do this in inline?

File: styles/Error.module.scss
Comment (spotandjake) [inline styles/Error.module.scss:1]: This should probably move to `styles/layout/Error.module.scss`

**Review by Kara-Zor-El — COMMENTED** [review]
Small changes

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]: Maybe we change this to: `An unknown error occured, with error code {statusCode}`, the client/server/code isnt very helpful to a user they just want to use our website. If you want you could add a button to direct them to the github issues page (If we do this maybe we make a specific template for that case).

File: pages/layout.tsx
Comment (spotandjake) [inline pages/layout.tsx]: After #6 is merged we need to add footer. (Maybe its not that simple, if it looks bad we might not keep).

File: layouts/Error.tsx
Comment (amberwelby) [inline layouts/Error.tsx]: Should this be pulled out and above main?

File: layouts/Error.tsx
Comment (Kara-Zor-El) [inline layouts/Error.tsx]: can we change this to our domain? (also, it should be either contains or include `https` in it)

File: layouts/Error.tsx
Comment (Kara-Zor-El) [inline layouts/Error.tsx]: this div and class is unnecessary, just put styles in `styles.page`

File: layouts/Error.tsx
Comment (Kara-Zor-El) [inline layouts/Error.tsx]: not exactly sure why we need a styling for the buttons when they already have a set style

File: layouts/Error.tsx
Comment (Kara-Zor-El) [inline layouts/Error.tsx]: lets try to avoid passing in children, instead pass in some props.
ex.
```ts
text: string,
image?: string
```

File: styles/layouts/Error.module.scss
Comment (Kara-Zor-El) [inline styles/layouts/Error.module.scss]: if we round the top, are we not gonna just have some whitespace at the very top of it? (when we get the new navbar)

File: styles/layouts/Error.module.scss
Comment (Kara-Zor-El) [inline styles/layouts/Error.module.scss]: even if we keep button, can be moved inside `.pageContent`

**Review by Kara-Zor-El — CHANGES_REQUESTED** [review]
Hey, great job.

Just some tiny nit-picks

File: layouts/Error.tsx
Comment (amberwelby) [reply layouts/Error.tsx]: Is this a best practices reason? I figured passing the child was the most customizable

File: layouts/Error.tsx
Comment (amberwelby) [reply layouts/Error.tsx]: Because the left is what it looked like without styling. And they didn't have a gap between them. I'll see where the cascading is messing it up 
image

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx:22]: Can you just add a comment below this, stating `Note that the Document. referrer property is not always reliable`.

I think the way this is implemented is the best option because we are a website most of the errors will be real 404s from mistypes, but we care about ones internally. but document refferer can be blocked by extensions, browsers and I think devices so its not 100% reliable.

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx]:
```suggestion
    if (document.referrer.includes('localhost:3000' || document.referrer.startsWith(departure_url))) {
```

I'm surprised the formatter did not fix that. Just a style nit though.

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx]:
```suggestion
    else { // If not, they don't need to report an issue
```

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx]:
```suggestion
```

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx]: Can we set `currentPage` to `''` I think it makes more sense to show no active page here.

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx:40]: instead of `stlyes.page` I feel like it might have made sense to take advantage of `scss` nesting here, considering we wont reuse the style. No need to change this just a small code thought.

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx]: Can we put children in a `div`, I think it makes sense to have uncontrollable elements wrapped in their own space rather in with our elements. It means we cant use grid or flex styles to seperate the elements.

File: layouts/Error.tsx
Comment (spotandjake) [reply layouts/Error.tsx]: Can you double check the default styles in the `button` module or if you need to add another style variant add it, I think instead of relying on external styling and cascading it might be better if we make the button component handle it internally.

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx]: instead of splitting this into two functions why not just inline the homeButton stuff and then do a simple:
```
if (allowReport) {
  <Button 
        type={ButtonType.LIGHT}
        href={reportHref}
        className={styles.button}
        label='Let us know'
      />
}
```
Alternatively just have the displayButtons. but inline the homeButton.

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx]: Why is this wrapped in a fragment?

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx]: Don't forget to fill this out.

File: pages/_error.tsx
Comment (spotandjake) [inline pages/_error.tsx]:
```suggestion
import { type NextPageContext } from 'next';
```
This should be a `type` import right?

File: styles/layouts/Error.module.scss
Comment (spotandjake) [inline styles/layouts/Error.module.scss:1]: Does this file use four space indents?

File: styles/layouts/Error.module.scss
Comment (spotandjake) [inline styles/layouts/Error.module.scss]: Can we not use `h-screen` here, My problem with it is, this layout now doesnt work on a page with where its expected to fit into a grid, I think on the parent screen we should handle a container thats the height we need and have this size to meet that, alternatively in the parent you can apply the `100vh` class if needed. 

Seperation of things like this means we can change other components without breaking things.

**Review by spotandjake — COMMENTED** [review]
Mostly looks good, A lot of style nits, and a few code layout changes I think would help simplify things, otherwise looks great.

File: pages/_error.tsx
Comment (amberwelby) [reply pages/_error.tsx]: I've seen both, I don't have a preference either way. 
https://nextjs.org/docs/pages/api-reference/functions/get-initial-props

Comment (spotandjake) [reply pages/_error.tsx]: > I've seen both, I don't have a preference either way. https://nextjs.org/docs/pages/api-reference/functions/get-initial-props

So the benefit of doing `type`is it allows for better treeshaking (This is mostly relevant when it relates to modules with sideeffects), it bassically tells typescript to extract the type info and not compile the values.

**Comment by spotandjake** [comment]
This pr needs to be rebased on main because of the changes in #24

**Comment by Kara-Zor-El** [comment]
> This pr needs to be rebased on main because of the changes in #24 

I don't think it uses the config

File: styles/layouts/Error.module.scss
Comment (Kara-Zor-El) [inline styles/layouts/Error.module.scss]: it might be slightly better to do 
```scss
.container {
  @apply flex flex-col;
  @apply size-full h-screen;
  @apply bg-color-3;

  .page {
    @apply flex flex-col flex-1;
    @apply p-4 gap-8;
    @apply size-full;
    @apply justify-center items-center text-center;
    @apply bg-color-1;
    @apply rounded-b-2xl;

    h1, p {
      @apply text-white;
    }

    h1 {
      @apply text-6xl font-bold;
    }

    .button{
      @apply m-2;
    }
  }
}
```
instead so that its nested more but that's a small nit

**Review by Kara-Zor-El — APPROVED** [review]
Looks great, small nit but ready for merging whenever you're ready

File: layouts/Error.tsx
Comment (spotandjake) [inline layouts/Error.tsx]: Can you run `npm run format` and `npm run lint` I don't think your formatter is working you can find the extension debug log in vscode under the output window (where the terminal is), and then prettier-eslint. (We can discuss in discord if its still giving problems).

**Review by spotandjake — APPROVED** [review]
One nit about formatting otherwise looks great. You have my approval (And we can merge once the formating is done), (Feel free to merge by yourself after that, use a squash comit).

**Pull request merged**
